helix-lsp/src/snippet.rs

```diff
@@ -368,23 +396,37 @@ mod parser {
         )
     }
 
-    fn anything<'a>() -> impl Parser<'a, Output = SnippetElement<'a>> {
-        // The parser has to be constructed lazily to avoid infinite opaque type recursion
-        |input: &'a str| {
-            let parser = choice!(tabstop(), placeholder(), choice(), variable());
+    fn anything<'a>(escape_chars: &'static [char]) -> impl Parser<'a, Output = SnippetElement<'a>> {
+        move |input: &'a str| {
+            let parser = choice!(
+                tabstop(),
+                placeholder(),
+                choice(),
+                variable(),
+                map(text(escape_chars), SnippetElement::Text)
+            );
             parser.parse(input)
         }
     }
 
     fn snippet<'a>() -> impl Parser<'a, Output = Snippet<'a>> {
-        let text = map(text(['$']), SnippetElement::Text);
-        map(one_or_more(choice!(anything(), text)), |parts| Snippet {
+        map(one_or_more(anything(TEXT_ESCAPE_CHARS)), |parts| Snippet {
             elements: parts,
         })
     }
 
-    pub fn parse(s: &str) -> Result<Snippet, &str> {
-        snippet().parse(s).map(|(_input, elements)| elements)
+    pub fn parse(s: &str) -> Result<(&str, Snippet), &str> {
+        snippet().parse(s)
+    }
+
+    pub fn parse_all(s: &str) -> Result<Snippet, &str> {
+        parse(s).and_then(|(remainder, snippet)| {
+            if remainder.is_empty() {
+                Ok(snippet)
+            } else {
+                Err(remainder)
+            }
+        })
```

Comment:
This seems like a reasonable behavior for `parse`. Can we just replace `parse` with the `parse_all` implementation?

Reply:
yeah, good point this makes the diff smaller too :+1:  I was thinking we might want to parse a snippet with something else afterward but we can just use `snippet().parse(text)` then